### PR TITLE
Added IPNS keys settings option

### PR DESCRIPTION
**Pull request by spylogsster**
Second part of https://github.com/brave/brave-browser/issues/15422
First part here https://github.com/brave/brave-core/pull/8609

- Added option to IPFS settings that links to keys management subpage, the option is visible only if local node is enabled
- Added APIs to add/remove/fetch keys from node, available only on settings page
- Added observers to switch UI when the node state changed (launch/shutdown)
- Added automatic redirect if current mode is not local node
- Added opportunity to launch node if it is stopped
- Added error indicator if node launch failed. Error message contains the link to open brave://ipfs page for error diagnostic

Normal mode
image

Node is not launched
image

Unable to start node
image

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Go to brave://ipfs and use option to add/remove keys
- The item to setup IPNS keys should be visible only if current method is local node and the node is launched
- Try to start/stop node when the add/remove key dialog is opened, the dialog should be closed automatically and the button to start node should be shown
- Try to open brave://settings/ipfs/keys page when the node is not installed. 
      - If current method is local node you should be asked to start the node and see launch error. There's a linked text that can be clicked to navigate to new brave://ipfs tab 
      - If current method is not local node you must be redirected to brave://ipfs
- if you start node from keys page, UIs block to start node and to add keys should be switched automatically

### CODE REVIEW

File: app/brave_generated_resources.grd [hunk 1]
```diff
@@ -757,6 +757,24 @@ By installing this extension, you are agreeing to the Google Widevine Terms of U
...
+        Setup your IPNS keys
+      </message>
+      <message name="IDS_SETTINGS_IPNS_KEYS_EDITOR_LINK_DESC" desc="The descriptoin for the link to open keys management block in IPFS settings">
+        Sign content by the corresponding private key and publish it to IPNS
+      </message>
+      <message name="IDS_SETTINGS_IPNS_KEYS_LIST_TITLE" desc="The title of the keys management list in IPFS settings">
+        Keys available for publishing
+      </message>
+      <message name="IDS_SETTINGS_IPNS_ADD_KEY_DIALOG_TITLE" desc="The title of the dialod to add new key on IPFS Settings page">
+        Enter name for new key
```
Comment (bbondy) [inline app/brave_generated_resources.grd:770]: Enter a name for the new key

File: app/brave_generated_resources.grd [hunk 2]
```diff
@@ -757,6 +757,24 @@ By installing this extension, you are agreeing to the Google Widevine Terms of U
...
+      </message>
+      <message name="IDS_SETTINGS_IPNS_ADD_KEY_DIALOG_ERROR" desc="The error message for the add key dialog on IPFS Settings page if key name is invalid">
+        This name can not be used
```
Comment (bbondy) [inline app/brave_generated_resources.grd:773]: This name cannot be used

File: app/brave_generated_resources.grd [hunk 3]
```diff
@@ -757,6 +757,24 @@ By installing this extension, you are agreeing to the Google Widevine Terms of U
       <message name="IDS_SETTINGS_RESOLVE_ENS_DESC" desc="The description for how to handle ENS">
         Method to resolve Ethereum Name Service (ENS)
       </message>
+      <message name="IDS_SETTINGS_IPNS_KEYS_EDITOR_LINK" desc="The title of the link to open keys management block in IPFS settings">
+        Setup your IPNS keys
+      </message>
+      <message name="IDS_SETTINGS_IPNS_KEYS_EDITOR_LINK_DESC" desc="The descriptoin for the link to open keys management block in IPFS settings">
```
Comment (bbondy) [inline app/brave_generated_resources.grd:763]: s/descriptoin/description

File: common/extensions/api/_ipfs_api_features.json [hunk 4]
```diff
@@ -54,5 +54,38 @@
     "contexts": ["blessed_extension"],
     "dependencies": [],
     "allowlist": ["780BF954C0F7C586EA9662D4F967771F49CC2114", "FF32507DC3DB5DFFD1D6733187C84D4B74713D63", "A321D47A2B4CA86898167A55CA8B2E02385EA7CD"]
+  }],
+  "ipfs.getIpnsKeysList": [{
+    "channel": "stable",
+    "contexts": ["blessed_extension"],
+    "dependencies": ["permission:ipfs"]
```
Comment (bbondy) [inline common/extensions/api/_ipfs_api_features.json:61]: I don't think we want to expose anything new unless we have a security review for it and need to for the chrome.ipfs API

**Review by bbondy — CHANGES_REQUESTED** [review]
Things that still need addressing:

- API Permissions fixing

- When testing it I noticed that when I reload brave://settings/ipfs and I click on:
`Setup your IPNS keys` it does a refresh on the page and doesn't work until I click it a second time.

- When I was testing it I clicked to add a key but got this in the console `Unchecked runtime.lastError: IPFS node is not launched`.  No error message was displayed though in the page.  I didn't have a node installed yet on my local node.  We should only show this section when local node is selected in preferences and a node is installed. 
-  After it is installed but I restart the browser so the node isn't running, I get `Unchecked runtime.lastError: Unable to create key`.  In this case I think it should start the already installed node.

**Comment by spylogsster** [comment]
> Things that still need addressing:
> 
> * API Permissions fixing
> * When testing it I noticed that when I reload brave://settings/ipfs and I click on:
>   `Setup your IPNS keys` it does a refresh on the page and doesn't work until I click it a second time.
> * When I was testing it I clicked to add a key but got this in the console `Unchecked runtime.lastError: IPFS node is not launched`.  No error message was displayed though in the page.  I didn't have a node installed yet on my local node.  We should only show this section when local node is selected in preferences and a node is installed.
> * After it is installed but I restart the browser so the node isn't running, I get `Unchecked runtime.lastError: Unable to create key`.  In this case I think it should start the already installed node.

done

**Comment by lidel** [comment]
Writing down  things from today's sync:
- Add Import and Export  buttons to load/save keypair files via `ipfs key import|export`
- Add ability to rotate `self` key via `ipfs key rotate`
    - `ipfs key rotate -o self.old.FOO`  may improve UX a bit (FOO could be lat 4 characters of old key)
    - `self` key rotation could be also added to `brave:ipfs` page as a privacy feature to "Rotate Peer Identity (PeerID)`

**Review by bbondy — APPROVED** [review]

**Pull request merged**